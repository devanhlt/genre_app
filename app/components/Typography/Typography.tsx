import { TxKeyPath, isRTL, translate } from "app/i18n"
import { appColors, typography } from "app/theme"
import i18n from "i18n-js"
import React from "react"
import { StyleProp, Text as RNText, TextProps as RNTextProps, TextStyle } from "react-native"

type Weights = keyof typeof typography.body02
type Presets = keyof typeof typography

export interface TypographyProps extends RNTextProps {
  /**
   * Text which is looked up via i18n.
   */
  tx?: TxKeyPath
  /**
   * The text to display if not using `tx` or nested components.
   */
  text?: string
  /**
   * Optional options to pass to i18n. Useful for interpolation
   * as well as explicitly setting locale or translation fallbacks.
   */
  txOptions?: i18n.TranslateOptions
  /**
   * An optional style override useful for padding & margin.
   */
  style?: StyleProp<TextStyle>
  /**
   * One of the different types of text presets.
   */
  preset?: Presets
  /**
   * Text weight modifier.
   */
  weight?: Weights

  /**
   * Color modifier.
   */
  color?: string
  /**
   * Children components.
   */
  children?: React.ReactNode
}

export function Typography(props: TypographyProps) {
  const {
    tx,
    txOptions,
    text,
    children,
    style: $styleOverride,
    color = appColors.common.characterPrimary,
    ...rest
  } = props

  const i18nText = tx && translate(tx, txOptions)
  const content = i18nText || text || children

  const preset: Presets = typography[props.preset] ? props.preset : "body02"

  const $styles = [$rtlStyle, { color }, typography[preset], $styleOverride]

  return (
    <RNText {...rest} style={$styles}>
      {content}
    </RNText>
  )
}

const $rtlStyle: TextStyle = isRTL ? { writingDirection: "rtl" } : {}
