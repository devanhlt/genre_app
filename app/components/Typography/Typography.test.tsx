/* eslint-disable react-native/no-inline-styles */
/* eslint-disable react-native/no-color-literals */
import React from "react"

import renderer from "react-test-renderer"
import { Typography } from "./Typography"

jest.mock("../../i18n/i18n", () => ({
  isRTL: true,
}))

describe("<Typography />", () => {
  it("renders correctly with isRTL is true", () => {
    const tree = renderer.create(
      <Typography preset="body01" text="body02 - Cillum eu laboris in labore" />,
    )
    expect(tree).toMatchSnapshot()
  })

  it("renders correctly with text which is looked up via i18n", () => {
    const Component = (
      <>
        <Typography preset="body02" text="body02 - Cillum eu laboris in labore" />

        <Typography preset="body03" text="body03 - Cillum eu laboris in labore" />

        <Typography preset="body04" text="body04 - Cillum eu laboris in labore" />

        <Typography preset="headline01" text="headline01 - Cillum eu laboris in labore" />

        <Typography preset="headline02" text="headline02 - Cillum eu laboris in labore" />

        <Typography preset="label01" text="label01 - Cillum eu laboris in labore" />

        <Typography preset="label02" text="label02 - Cillum eu laboris in labore" />

        <Typography preset="label03" text="label03 - Cillum eu laboris in labore" />

        <Typography preset="link01" text="link01 - Cillum eu laboris in labore" />

        <Typography preset="link02" text="link02 - Cillum eu laboris in labore" />

        <Typography preset="support01" text="support01 - Cillum eu laboris in labore" />

        <Typography preset="support02" text="support02 - Cillum eu laboris in labore" />

        <Typography preset="title01" text="title01 - Cillum eu laboris in labore" />

        <Typography preset="title02" tx="common.back" />

        <Typography preset="title02">
          <Typography preset="title02" tx="common.back" />
        </Typography>
      </>
    )
    const tree = renderer.create(Component)
    expect(tree).toMatchSnapshot()
  })
})
