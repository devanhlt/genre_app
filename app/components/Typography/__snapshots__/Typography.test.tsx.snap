// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Typography /> renders correctly with isRTL is true 1`] = `
<Text
  preset="body01"
  style={
    [
      {
        "writingDirection": "rtl",
      },
      {
        "color": "#000F27",
      },
      {
        "fontFamily": "System",
        "fontSize": 14,
        "fontWeight": "400",
        "letterSpacing": 0.15,
        "lineHeight": 20,
      },
      undefined,
    ]
  }
>
  body02 - Cillum eu laboris in labore
</Text>
`;

exports[`<Typography /> renders correctly with text which is looked up via i18n 1`] = `
[
  <Text
    preset="body02"
    style={
      [
        {
          "writingDirection": "rtl",
        },
        {
          "color": "#000F27",
        },
        {
          "fontFamily": "System",
          "fontSize": 16,
          "fontWeight": "400",
          "letterSpacing": 0.15,
          "lineHeight": 24,
        },
        undefined,
      ]
    }
  >
    body02 - Cillum eu laboris in labore
  </Text>,
  <Text
    preset="body03"
    style={
      [
        {
          "writingDirection": "rtl",
        },
        {
          "color": "#000F27",
        },
        {
          "fontFamily": "System",
          "fontSize": 14,
          "fontWeight": "600",
          "letterSpacing": 0.15,
          "lineHeight": 20,
        },
        undefined,
      ]
    }
  >
    body03 - Cillum eu laboris in labore
  </Text>,
  <Text
    preset="body04"
    style={
      [
        {
          "writingDirection": "rtl",
        },
        {
          "color": "#000F27",
        },
        {
          "fontFamily": "System",
          "fontSize": 16,
          "fontWeight": "600",
          "letterSpacing": 0.15,
          "lineHeight": 24,
        },
        undefined,
      ]
    }
  >
    body04 - Cillum eu laboris in labore
  </Text>,
  <Text
    preset="headline01"
    style={
      [
        {
          "writingDirection": "rtl",
        },
        {
          "color": "#000F27",
        },
        {
          "fontFamily": "System",
          "fontSize": 24,
          "fontWeight": "600",
          "letterSpacing": 0,
          "lineHeight": 32,
        },
        undefined,
      ]
    }
  >
    headline01 - Cillum eu laboris in labore
  </Text>,
  <Text
    preset="headline02"
    style={
      [
        {
          "writingDirection": "rtl",
        },
        {
          "color": "#000F27",
        },
        {
          "fontFamily": "System",
          "fontSize": 20,
          "fontWeight": "600",
          "letterSpacing": 0,
          "lineHeight": 28,
        },
        undefined,
      ]
    }
  >
    headline02 - Cillum eu laboris in labore
  </Text>,
  <Text
    preset="label01"
    style={
      [
        {
          "writingDirection": "rtl",
        },
        {
          "color": "#000F27",
        },
        {
          "fontFamily": "System",
          "fontSize": 14,
          "fontWeight": "600",
          "letterSpacing": 0.15,
          "lineHeight": 20,
        },
        undefined,
      ]
    }
  >
    label01 - Cillum eu laboris in labore
  </Text>,
  <Text
    preset="label02"
    style={
      [
        {
          "writingDirection": "rtl",
        },
        {
          "color": "#000F27",
        },
        {
          "fontFamily": "System",
          "fontSize": 16,
          "fontWeight": "600",
          "letterSpacing": 0.15,
          "lineHeight": 24,
        },
        undefined,
      ]
    }
  >
    label02 - Cillum eu laboris in labore
  </Text>,
  <Text
    preset="label03"
    style={
      [
        {
          "writingDirection": "rtl",
        },
        {
          "color": "#000F27",
        },
        {
          "fontFamily": "System",
          "fontSize": 14,
          "fontWeight": "400",
          "letterSpacing": 0.15,
          "lineHeight": 20,
        },
        undefined,
      ]
    }
  >
    label03 - Cillum eu laboris in labore
  </Text>,
  <Text
    preset="link01"
    style={
      [
        {
          "writingDirection": "rtl",
        },
        {
          "color": "#000F27",
        },
        {
          "fontFamily": "System",
          "fontSize": 14,
          "fontWeight": "500",
          "letterSpacing": 0.15,
          "lineHeight": 20,
        },
        undefined,
      ]
    }
  >
    link01 - Cillum eu laboris in labore
  </Text>,
  <Text
    preset="link02"
    style={
      [
        {
          "writingDirection": "rtl",
        },
        {
          "color": "#000F27",
        },
        {
          "fontFamily": "System",
          "fontSize": 16,
          "fontWeight": "500",
          "letterSpacing": 0.15,
          "lineHeight": 24,
        },
        undefined,
      ]
    }
  >
    link02 - Cillum eu laboris in labore
  </Text>,
  <Text
    preset="support01"
    style={
      [
        {
          "writingDirection": "rtl",
        },
        {
          "color": "#000F27",
        },
        {
          "fontFamily": "System",
          "fontSize": 12,
          "fontWeight": "400",
          "letterSpacing": 0.25,
          "lineHeight": 16,
        },
        undefined,
      ]
    }
  >
    support01 - Cillum eu laboris in labore
  </Text>,
  <Text
    preset="support02"
    style={
      [
        {
          "writingDirection": "rtl",
        },
        {
          "color": "#000F27",
        },
        {
          "fontFamily": "System",
          "fontSize": 12,
          "fontWeight": "600",
          "letterSpacing": 0.25,
          "lineHeight": 16,
        },
        undefined,
      ]
    }
  >
    support02 - Cillum eu laboris in labore
  </Text>,
  <Text
    preset="title01"
    style={
      [
        {
          "writingDirection": "rtl",
        },
        {
          "color": "#000F27",
        },
        {
          "fontFamily": "System",
          "fontSize": 24,
          "fontWeight": "600",
          "letterSpacing": 0.1,
          "lineHeight": 32,
        },
        undefined,
      ]
    }
  >
    title01 - Cillum eu laboris in labore
  </Text>,
  <Text
    preset="title02"
    style={
      [
        {
          "writingDirection": "rtl",
        },
        {
          "color": "#000F27",
        },
        {
          "fontFamily": "System",
          "fontSize": 20,
          "fontWeight": "600",
          "letterSpacing": 0.1,
          "lineHeight": 28,
        },
        undefined,
      ]
    }
  >
    common.back undefined
  </Text>,
  <Text
    preset="title02"
    style={
      [
        {
          "writingDirection": "rtl",
        },
        {
          "color": "#000F27",
        },
        {
          "fontFamily": "System",
          "fontSize": 20,
          "fontWeight": "600",
          "letterSpacing": 0.1,
          "lineHeight": 28,
        },
        undefined,
      ]
    }
  >
    <Text
      preset="title02"
      style={
        [
          {
            "writingDirection": "rtl",
          },
          {
            "color": "#000F27",
          },
          {
            "fontFamily": "System",
            "fontSize": 20,
            "fontWeight": "600",
            "letterSpacing": 0.1,
            "lineHeight": 28,
          },
          undefined,
        ]
      }
    >
      common.back undefined
    </Text>
  </Text>,
]
`;
